notes/stv/20200429-apache-arrow.txt

```diff
@@ -0,0 +1,128 @@
+    )
+
+pandas_df = df.select("*").toPandas()
+pandas_df.head()
+
+
+# Exception triggered:
+
+     >  Job aborted due to stage failure: Task 3901 in stage 15.0 failed 4 times, most recent failure: Lost task 3901.3 in stage 15.0 (TID 41499, stv-dev-worker-8, executor 266): TaskResultLost (result lost from block manager)
+
+
+
+# Let's try to see if we can improve the performance with Apache Arrow
+
+
+
+# Installing Apache Arrow
+# -------------------------------
+
+
+https://docs.databricks.com/spark/latest/spark-sql/spark-pandas.html
+https://arrow.apache.org/install/
+
+
+Apache Arrow is an in-memory columnar data format used in Apache Spark to efficiently transfer data between JVM and Python processes. This is beneficial to Python developers that work with pandas and NumPy data. However, its usage is not automatic and requires some minor changes to configuration or code to take full advantage and ensure compatibility.
+
+Convert Spark DataFrames to and from pandas DataFrames
+Arrow is available as an optimization when converting a Spark DataFrame to a pandas DataFrame using the call toPandas() and when creating a Spark DataFrame from a pandas DataFrame with createDataFrame(pandas_df). To use Arrow when executing these calls, set the Spark configuration spark.sql.execution.arrow.enabled to true. This configuration is disabled by default.
+
+
+
+# Repeat on each worker node
+# user@worker
+# -------------------------------
+
+pip install pyarrow==0.17.*
+
+
+# Try the following Spark job (Via Zeppelin/Pyspark)
+# user@worker
+# -------------------------------
+
+
+%spark.pyspark
+import pandas as pd
+import numpy as np
+
+# Enable Arrow-based columnar data transfers
+spark.conf.set("spark.sql.execution.arrow.enabled", "true")
+
+# Generate a pandas DataFrame
+pdf = pd.DataFrame(np.random.rand(100, 3))
+
+# Create a Spark DataFrame from a pandas DataFrame using Arrow
+df = spark.createDataFrame(pdf)
+
+# Convert the Spark DataFrame back to a pandas DataFrame using Arrow
+result_pdf = df.select("*").toPandas()
+
+
+/home/fedora/spark/python/lib/pyspark.zip/pyspark/sql/session.py:714: UserWarning: createDataFrame attempted Arrow optimization because 'spark.sql.execution.arrow.enabled' is set to true; however, failed by the reason below:
+  'JavaPackage' object is not callable
+Attempting non-optimization as 'spark.sql.execution.arrow.fallback.enabled' is set to true.
+
+
+
+# After repeating the cell run, the error disappears..(!?)
+
+
+
+# Selecting the above gaia query dataframe into a pandas object
+# user@worker
+# -------------------------------
+# Let's Try running the following Zeppelin cell from the notebook example:
+
+%spark.pyspark
+#df = pd.DataFrame(df, columns=df.columns)
+#print(type(df))
+#df.head()
+# ... doesn't work for me for the system install of pandas. Try this:
+pandas_df = df.select("*").toPandas()
+print(type(pandas_df))
+pandas_df.head()
+
+
```

Comment:
So in summary -
1.  Ran the SQL query at line 36-42 to create a Spark data frame, and got an exception when trying to convert the Spark data frame to a Pandas data frame.
2. Install Apache Arrow on _all_ of the worker nodes
3.  Ran the **same** SQL query at line 36-42 to create a Spark data frame, and converting the Spark data frame to a Pandas data frame now works ?